daemon/container_operations.go

```diff
@@ -574,52 +574,86 @@ func hasUserDefinedIPAddress(ipamConfig *networktypes.EndpointIPAMConfig) bool {
 	return ipamConfig != nil && (len(ipamConfig.IPv4Address) > 0 || len(ipamConfig.IPv6Address) > 0)
 }
 
-// User specified ip address is acceptable only for networks with user specified subnets.
-func validateNetworkingConfig(n *libnetwork.Network, epConfig *networktypes.EndpointSettings) error {
-	if n == nil || epConfig == nil {
+// validateEndpointSettings checks whether the given epConfig is valid. The nw parameter might be nil as a container
+// can be created with a reference to a network that don't exist yet. In that case, only partial validation will be
+// done.
+func validateEndpointSettings(nw *libnetwork.Network, nwName string, epConfig *networktypes.EndpointSettings) error {
+	if epConfig == nil {
 		return nil
 	}
-	if !containertypes.NetworkMode(n.Name()).IsUserDefined() {
+
+	var errs []error
+	if !containertypes.NetworkMode(nwName).IsUserDefined() {
+		// On Linux, user specified IP address is accepted only for networks with user specified subnets.
 		if hasUserDefinedIPAddress(epConfig.IPAMConfig) && !enableIPOnPredefinedNetwork() {
-			return runconfig.ErrUnsupportedNetworkAndIP
+			errs = append(errs, runconfig.ErrUnsupportedNetworkAndIP)
 		}
 		if len(epConfig.Aliases) > 0 && !serviceDiscoveryOnDefaultNetwork() {
-			return runconfig.ErrUnsupportedNetworkAndAlias
+			errs = append(errs, runconfig.ErrUnsupportedNetworkAndAlias)
 		}
 	}
+
 	if !hasUserDefinedIPAddress(epConfig.IPAMConfig) {
+		return errors.Join(errs...)
+	}
+	if epConfig.IPAMConfig.IPv4Address != "" && net.ParseIP(epConfig.IPAMConfig.IPv4Address).To4() == nil {
```

Comment:
Looks like a bunch of these are also done in `buildCreateEndpointOptions`. Wondering if we can extract some of these checks, and make both use the same validation 🤔 https://github.com/moby/moby/blob/220fba06e7ebd6798f24fa690ba9069067ecea86/daemon/network.go#L808-L838

Reply:
As shared privately, for now we can't really drop the validation steps in `buildCreateEndpointOptions`. We first need to make some refactorings to make sure user-supplied data are correctly sanitized and transformed into useful structs instead of keeping literals everywhere.

This will be addressed in a follow-up PR.